### PR TITLE
[Harvest] Handle cipher selection

### PR DESCRIPTION
**Pull request by drazik**
When the user clicks on a cipher, it is used in the account form.

There are two cases:

* If all required fields are present in the cipher, then the account form is automatically submitted
* Otherwise, the user can edit the missing fields and submit the form manually

### CODE REVIEW

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 1]
```diff
@@ -131,11 +135,111 @@ export class TriggerManager extends Component {
...
+    cipherData.login.username = login
+    cipherData.login.password = password
+
+    const cipher = await vaultClient.createNewCipher(cipherData, originalCipher)
+    await vaultClient.saveCipher(cipher)
+
+    return cipher.id
+  }
+
+  async getCipherId(login, password) {
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: We can remove this function now.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 2]
```diff
@@ -226,10 +299,38 @@ export class TriggerManager extends Component {
...
+    const { fields } = konnector
+    const sanitizedFields = manifest.sanitizeFields(fields)
+    const account = this.cipherToAccount(selectedCipher)
+    const values = manifest.getDefaultedValues(konnector, account)
+
+    const requiredFields = Object.entries(sanitizedFields)
+      .filter(([, value]) => value.required)
+      .map(([key]) => key)
+
+    const requiredFieldsInValues = intersection(
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: How about `const hasValuesForRequiredFields = intersction().length === requiredFields.length`, and then use `hasValuesForRequiredFields` in the condition below?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 3]
```diff
@@ -276,11 +400,13 @@ export class TriggerManager extends Component {
           )}
           {step === 'accountForm' && (
             <>
-              <BackButton onClick={this.showCiphersList}>
-                {t('triggerManager.backToCiphersList')}
-              </BackButton>
+              {!account && (
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Maybe we can be more explicit here by introducing a `isUpdate = !account` variable?

File: packages/cozy-harvest-lib/src/helpers/manifest.js [hunk 4]
```diff
@@ -193,9 +193,20 @@ export const sanitize = (manifest = {}) =>
       }
     : manifest
 
+const getDefaultedValues = (konnector, account) => {
```
Comment (y-lohse) [inline packages/cozy-harvest-lib/src/helpers/manifest.js]: How about `getFieldValues` ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 5]
```diff
@@ -131,11 +135,111 @@ export class TriggerManager extends Component {
       this.handleError(error)
     }
   }
+
+  getSelectedCipherId() {
+    const { selectedCipher } = this.state
+    return selectedCipher && selectedCipher.id
+  }
+
+  async getNewCipherId(login, password) {
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: can you add the jsdoc ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 6]
```diff
@@ -131,11 +135,111 @@ export class TriggerManager extends Component {
...
+    const konnectorName = get(konnector, 'name') || get(konnector, 'slug')
+
+    const cipherData = {
+      id: null,
+      type: 1,
+      name: konnectorName,
+      login: {
+        username: login,
+        password,
+        uris: konnectorURI ? [{ uri: konnectorURI, match: 0 }] : []
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: what is match:0 ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 7]
```diff
@@ -238,6 +339,29 @@ export class TriggerManager extends Component {
...
+      auth: pickBy(
+        {
+          login: cipher.login.username,
+          password: cipher.login.password,
+          ...customFields
+        },
+        value => Boolean(value)
+      )
+    }
+  }
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Should we create a `Cipher` helper / doctype?

File: packages/cozy-harvest-lib/src/helpers/manifest.js (hunk 4 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/helpers/manifest.js]: Yep I like it, thanks

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 6 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Maybe @y-lohse or @edas have the answer? I just moved this part of the code in a refactor, and have no idea why we have this `match: 0` here actually.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 7 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Or this could be part of the `Account` one? Like `Account::fromCipher`?

Comment (Crash--) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Deal :)

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 8]
```diff
@@ -131,11 +135,85 @@ export class TriggerManager extends Component {
       this.handleError(error)
     }
   }
+
+  /**
+   * Get the ID of the cipher selected by the user in the list
+   *
+   * @returns {string} the cipher ID
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: false | string

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 6 above)
Comment (y-lohse) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Right, I missed that. It's the type of `uri` that we are using — full URL match, regular expression, etc. It's an ENUM, `0` presumably means "Full URL match" (@edas can you confirm?).

I think we can replace this with a more meaningful constant.

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 9]
```diff
@@ -131,11 +135,85 @@ export class TriggerManager extends Component {
...
+
+  /**
+   * Find an existing cipher for the account and return its ID
+   *
+   * @param {string} login - the login to set in the cipher
+   * @param {string} password - the password to set in the cipher
+   *
+   * @returns {string} the cipher ID
+   */
+  async getExistingCipherIdForAccount(login, password) {
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Should we pass the `account` in params? and move this method to the account model? 

Also, we say `Find an existing cipher` but it seems that we create a `newCypher` 🤔

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 6 above)
Comment (edas) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: from the jslib/enums/uriMatchType.ts: 

```typescript
export enum UriMatchType {
    Domain = 0,
    Host = 1,
    StartsWith = 2,
    Exact = 3,
    RegularExpression = 4,
    Never = 5,
}
``` 

0 means "match only with the domain part", which is the default.

Agree that we should probably have a constant here. I can export the enum from cozy-keys-lib so you can reuse it. Would it make sense ?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 10]
```diff
@@ -131,11 +135,111 @@ export class TriggerManager extends Component {
...
+
+  async getNewCipherId(login, password) {
+    const { vaultClient, konnector } = this.props
+
+    const konnectorURI = get(konnector, 'vendor_link')
+    const konnectorName = get(konnector, 'name') || get(konnector, 'slug')
+
+    const cipherData = {
+      id: null,
+      type: 1,
```
Comment (edas) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]:
```suggestion
      type: CipherType.Login,
```

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx [hunk 11]
```diff
@@ -226,10 +273,37 @@ export class TriggerManager extends Component {
...
+    const account = this.cipherToAccount(selectedCipher)
+    const values = manifest.getFieldsValues(konnector, account)
+
+    const requiredFields = Object.entries(sanitizedFields)
+      .filter(([, value]) => value.required)
+      .map(([key]) => key)
+
+    const hasValuesForRequiredFields =
+      intersection(Object.keys(values), requiredFields).length ===
+      requiredFields.length
```
Comment (Crash--) [inline packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Should we move those previous lines to an helper? It seems that they are not a React related stuff, but more "konnector/account"?

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 9 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Actually the `createCipher` seems not to be so well-named. It's more like a `createOrUpdate`

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 7 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: See https://github.com/cozy/cozy-libs/pull/782/commits/712456d95d2e461054896280a139ce713d7403f1

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 6 above)
Comment (edas) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: In https://github.com/cozy/cozy-keys-lib/pull/25

File: packages/cozy-harvest-lib/src/components/TriggerManager.jsx (hunk 11 above)
Comment (drazik) [reply packages/cozy-harvest-lib/src/components/TriggerManager.jsx]: Done in https://github.com/cozy/cozy-libs/pull/782/commits/5a29d9fb712937e6095b2ce0b2d92a86f6aee79f

File: packages/cozy-doctypes/src/Account.js [hunk 12]
```diff
@@ -29,6 +30,36 @@ class Account extends Document {
...
+    const customFields = (cipher.fields || []).reduce((fields, field) => {
+      fields[field.name] = field.value
+
+      return fields
+    }, {})
+
+    return {
+      auth: pickBy(
+        {
+          login: cipher.login.username,
```
Comment (edas) [inline packages/cozy-doctypes/src/Account.js]: Il manque le relationships avec le cipher justement

Comment (drazik) [reply packages/cozy-doctypes/src/Account.js]: En fait cette méthode sert à transformer le Cipher en `io.cozy.accounts` compatible avec le composant `AccountForm`. Dans ce cas on n'a besoin que de la propriété `auth`. L'ajout de la relationship est faite par un helper [`setVaultCipherRelationship`](https://github.com/cozy/cozy-libs/pull/782/files#diff-436ec256172e55116b84d1c18c97218aR248), qui pourrait effectivement être déplacé dans cozy-doctypes aussi.

Comment (edas) [reply packages/cozy-doctypes/src/Account.js]: Ok mais là tu fais une méthode avec un nom générique dans une classe utilitaire générique, et elle renvoie un contenu incomplet. Du coup soit tu complètes soit tu mets un nom qui correspond mieux

**Pull request merged**